Implement phone number analyzer

rursprung opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
this is largely based on [elasticsearch-phone] and internally uses
[libphonenumber].
this intentionally only ports a subset of the features: only `phone` and
`phone-search` are supported right now, `phone-email` can be added
if/when there's a clear need for it.

using `libphonenumber` is required since parsing phone numbers is a
non-trivial task (even though it might seem trivial at first glance!),
as can be seen in the list [falsehoods programmers believe about phone
numbers][falsehoods].

this allows defining the region to be used when analysing a phone
number. so far only the generic "unkown" region (`ZZ`) had been used
which worked as long as international numbers were prefixed with `+` but
did not work when using local numbers (e.g. a number stored as
`+4158...` was not matched against a number entered as `004158...` or
`058...`).

example configuration for an index:
```json
{
  "index": {
    "analysis": {
      "analyzer": {
        "phone": {
          "type": "phone"
        },
        "phone-search": {
          "type": "phone-search"
        },
        "phone-ch": {
          "type": "phone",
          "phone-region": "CH"
        },
        "phone-search-ch": {
          "type": "phone-search",
          "phone-region": "CH"
        }
      }
    }
  }
}
```
this creates four analyzers: `phone` and `phone-search` which do not
explicitly specify a region and thus fall back to `ZZ` (unknown region,
regional version of international dialing prefix (e.g. `00` instead of
`+` in most of europe) will not be recognised) and `phone-ch` and
`phone-search-ch` which will try to parse the phone number as a swiss
phone number (thus e.g. `00` as a prefix is recognised as the
international dialing prefix).

note that the analyzer is (currently) not meant to find phone numbers in
large text documents - instead it should be used on fields which contain
just the phone number (though extra text will be ignored) and it
collects the whole content of the field into a `String` in memory,
making it unsuitable for large field values.

this has been implemented in a new plugin which is however part of the
central opensearch repository as it was deemed too big an overhead to
have it in a separate repository but not important enough to bundle it
directly in `analysis-common` (see the discussion on the issue and the
PR for further details).

note that the new plugin has been added to the exclude list of the
javadoc check as this check is overzealous and also complains in many
cases where it shouldn't (e.g. on overridden methods - which it should
theoretically not do - or constructors which don't even exist). the
check first needs to be improved before this exclusion could be removed.

closes opensearch-project/OpenSearch#11326

[elasticsearch-phone]: https://github.com/purecloudlabs/elasticsearch-phone
[libphonenumber]: https://github.com/google/libphonenumber
[falsehoods]: https://github.com/google/libphonenumber/blob/master/FALSEHOODS.md

Signed-off-by: Ralph Ursprung <Ralph.Ursprung@avaloq.com>

### Related Issues
Resolves #11326

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable. => https://github.com/opensearch-project/documentation-website/issues/8389

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

rursprung commented:
> testShardStoreFetchMultiNodeMultiIndexesUsingBatchAction



> ❕ Gradle check result for [f7669e2](https://github.com/opensearch-project/OpenSearch/commit/f7669e2e732ea1a5a1936f07723671a4d3f39238): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/47800/)
> 
>     * **TEST FAILURES:**
> 
> 
> ```
>       1 org.opensearch.gateway.RecoveryFromGatewayIT.testShardStoreFetchMultiNodeMultiIndexesUsingBatchAction
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

this is a flaky test: #14304

rursprung commented:
this should now be ready for review :rocket:
CC a couple of people who were involved in the discussions (in no particular order):
@reta (for the things we discussed on slack about `Strings` & `Streams`)
@msfroh (i now ended up not using `RemoveDuplicatesTokenFilter` as using a `Set` internally remains much simpler and means that the tokenizer already returns unique tokens plus i ran into the error `first position increment must be > 0 (got 0) for field 'phone'` while trying to get it to work)
@dblock
@timsmithgenesys (as discussed in #11326 this is based on the old ES plugin from your company - thanks again for licensing it as Apache 2.0! ❤️)
@macohen (originally pointed out the old ES plugin)

dblock commented:
Thanks! Probably @msfroh is our best reviewer for this.

reta commented on libs/core/src/main/java/org/opensearch/core/common/Strings.java:
```diff
@@ -815,4 +815,17 @@ public static String toLowercaseAscii(String in) {
+     * @param s the string, must not be null.
+     * @return true if the string is empty or only contains digits, false otherwise.
+     */
+    public static boolean isDigits(final String s) {
```
> * Important: this also returns {@code true} for empty strings!

I think this is very surprising  behaviour  and should be revisited, the empty string is not valid set of digits.

reta reviewed (changes requested):
Thanks @rursprung , I feel strongly that we should have it under separate plugin (`analsys-phonenumber` fe) and keep the `analysis-common` clean (not everyone need this functionality)

rursprung commented:
> Thanks @rursprung , I feel strongly that we should have it under separate plugin (`analysis-phonenumber` fe) and keep the `analysis-common` clean (not everyone needs this functionality)

@reta: i put it in `analysis-common` based on @msfroh's comment here: https://github.com/opensearch-project/OpenSearch/issues/11326#issuecomment-2334939361 (seconded also by @dblock). i'm happy to move things around if this is the general consensus but would like to avoid doing that a few more times :)

msfroh commented:
@reta -- do you mean a separate plug-in in a separate repo? Or just a separate plug-in in the core repo?

In my opinion, the overhead of maintaining a separate plugin build/distribution is not worth it for a single analyzer class of ~20 lines.

reta commented:
> separate plug-in in the core repo?

Thanks @rursprung , yes, I mean a separate plug-in in the core repo (under `plugins`), it is indeed small enough.

reta commented:
> @reta: i put it in `analysis-common` based on @msfroh's comment here: [#11326 (comment)](https://github.com/opensearch-project/OpenSearch/issues/11326#issuecomment-2334939361) (seconded also by @dblock). i'm happy to move things around if this is the general consensus but would like to avoid doing that a few more times :)

My apologies @rursprung , I missed that but would have commented there. If folks think `analysis-common` is the place - no objection (but in my option, it does not belong here, the phone numbers are relevant to small percentage of the applications and is not common case).

msfroh commented:
Between analysis-common and a separate plugin in this repo, I don't have strong opinions. I definitely agree that it's not likely to be widely used, so a separate plugin in this repo gives people the choice to install it if they want it.

rursprung commented:
wouldn't the separate plugin in this repo cause quite a bit of overhead for the release process? (i'm not familiar with that part but would presume that it'd have to be patched in at several places to make sure that it gets published?)

having it in `analysis-common` would have one major benefit for me: it'd become available on all public cloud offerings (once they upgrade to that release) - and i need it on at least some of them 🙂.
also, i don't think that it's a less common use-case than many other analyzers currently in the `analysis-common` package and my gut feeling (without having done a market survey) is that it'll be used once it's available (and communicated) since there are many use-cases where you're ingesting phone numbers, e.g. as part of contact data (true, you might not need this for a web shop).

reta commented:
> wouldn't the separate plugin in this repo cause quite a bit of overhead for the release process? (i'm not familiar with that part but would presume that it'd have to be patched in at several places to make sure that it gets published?)

No, we build and release the core as a whole

> having it in analysis-common would have one major benefit for me: it'd become available on all public cloud offerings (once they upgrade to that release) - and i need it on at least some of them 🙂.

It would be the same plugin as any other bundled with core and installable with `opensearch-plugin` tool. `analysis-common` is a module which means you cannot install it, what it or not. For example, assume there is CVE published against `libphonenumber`. The mitigation with plugin is easy - uninstall it till fix available, with module it is impossible.

rursprung commented:
> It would be the same plugin as any other bundled with core and installable with `opensearch-plugin` tool. `analysis-common` is a module which means you cannot uninstall it, want it or not. For example, assume there is CVE published against `libphonenumber`. The mitigation with plugin is easy - uninstall it till fix available, with module it is impossible.

you say "bundled" - does that mean that it'll be part of the standard (or even `min`?) opensearch distribution by default if it's a plugin in this repo and it can just optionally be uninstalled rather than being something which has to be installed explicitly?

reta commented:
> it's a plugin in this repo and it can just optionally be uninstalled rather than being something which has to be installed explicitly?

it would be a plugin which is **not installed** by default, has to be installed and uinstalled explicitly

rursprung commented:
> it would be a plugin which is **not installed** by default, has to be installed and uninstalled explicitly when there is a need to use it.

oh 🙁
well, i guess it's the same as an external plugin with the advantage that it's in this repo, making maintenance easier (releasing & publishing happens out of the box).

so, is it the final consensus of the reviewers that i should move it to a new `analysis-phonenumber` plugin within this repo? if so, i'll take care of it next week

reta commented:
> so, is it the final consensus of the reviewers that i should move it to a new `analysis-phonenumber` plugin within this repo? if so, i'll take care of it next week

+1 from me, thanks @rursprung

rursprung commented:
i've done the move and now just need to make some checks happy (namely the javadoc one; it seems to ignore some things in common but not here) and give it another manual test spin and then we should be good to go. just pushed it so that you can have a first look as i don't expect to change the business logic of it anymore (hasn't changed since the last push either)

reta commented:
> i've done the move and now just need to make some checks happy (namely the javadoc one; it seems to ignore some things in common but not here) and give it another manual test spin and then we should be good to go. just pushed it so that you can have a first look as i don't expect to change the business logic of it anymore (hasn't changed since the last push either)

Thanks @rursprung , the plugin should be under `plugins` folder, the modules are plugins that are installed by default, As far as I remember, we concluded that this plugin is optional (opt-in).

reta commented:
It looks great, thanks @rursprung a few things left:
 - could you please add YAML tests for analyzer itself (fe `analysis-icu/src/yamlRestTest/resources/rest-api-spec/test/analysis_icu/20_search.yml`)
 - please create a documentation issue for 2.18.0 release to document this new analyzer (https://github.com/opensearch-project/documentation-website/)

Thank you.

rursprung commented:
> namely the javadoc one; it seems to ignore some things in common but not here

turns out that check is excluded for most existing plugins, so that's why it didn't fail there 😅:
https://github.com/opensearch-project/OpenSearch/blob/e8b02c9545d7908ae1a3ea2d109c06ff3e0c07a5/gradle/missing-javadoc.gradle#L109

as discussed on slack i've now added the new plugin to that list as well as the check is overzealous (details also in the commit-msg)

> Thanks @rursprung , the plugin should be under `plugins` folder, the modules are plugins that are installed by default, As far as I remember, we concluded that this plugin is optional (opt-in).

ah, i didn't spot that yet. moved it now.

should now be ready for review! (the previous failure seems like a false reject; i just rebased now, hopefully it goes away by itself)

rursprung commented:
sorry, missed your comment before :(

> * could you please add YAML tests for analyzer itself (fe `analysis-icu/src/yamlRestTest/resources/rest-api-spec/test/analysis_icu/20_search.yml`)

does this add any benefit over the java-based test? or should i even replace the java-based test with this one?



> * please create a documentation issue for 2.18.0 release to document this new analyzer (https://github.com/opensearch-project/documentation-website/)

if everything (names of analyzers, index setting, etc.) is ok for you then i can do this (just wanted to avoid having to get it updated several times)

reta commented:
> does this add any benefit over the java-based test? or should i even replace the java-based test with this one?

it gives a good example of how the analyzer could be used (+ very consistent with the test anaylsis-* suites)

> if everything (names of analyzers, index setting, etc.) is ok for you then i can do this (just wanted to avoid having to get it updated several times)

Sure, this is just an issue (not a documentation update), so docs team knows that new features will be added to the release, there is no need to create a pull request now

rursprung commented:
> ❌ Gradle check result for [f1dffe9](https://github.com/opensearch-project/OpenSearch/commit/f1dffe95dee29d1471fd2ce0852af5e59b54f713): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48526/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

flaky test: https://github.com/opensearch-project/OpenSearch/issues/16109



> > does this add any benefit over the java-based test? or should i even replace the java-based test with this one?
> 
> it gives a good example of how the analyzer could be used (+ very consistent with the test anaylsis-* suites)

i'll update it next week

> > if everything (names of analyzers, index setting, etc.) is ok for you then i can do this (just wanted to avoid having to get it updated several times)
> 
> Sure, this is just an issue (not a documentation update), so docs team knows that new features will be added to the release, there is no need to create a pull request now

done: https://github.com/opensearch-project/documentation-website/issues/8389

rursprung commented:
> > > does this add any benefit over the java-based test? or should i even replace the java-based test with this one?
> > 
> > 
> > it gives a good example of how the analyzer could be used (+ very consistent with the test anaylsis-* suites)
> 
> i'll update it next week

i've now added the test.

note that the `analysis-icu` example is probably not particularly good since it's testing with a french text, thus not testing the ICU part at all 😅

reta commented:
@rursprung thank you, could you please resolve the conflicts?

reta commented on plugins/analysis-phonenumber/src/main/java/org/opensearch/analysis/phone/PhoneNumberAnalysisPlugin.java:
```diff
@@ -0,0 +1,60 @@
+public class PhoneNumberAnalysisPlugin extends Plugin implements AnalysisPlugin {
+
+    @Override
+    public Map<String, AnalysisModule.AnalysisProvider<AnalyzerProvider<? extends Analyzer>>> getAnalyzers() {
```
This is just a question: do you think we will be better off with `phone-ngrams` and `phone` names for analyzers?  The `phone-search` is somewhat confusing (at least to me) since in both cases the phone could be searched, right?

rursprung replied:
i followed the pattern of the old ES plugin (but nobody is using that anymore, so no preservation of vested rights needed there).
the reasoning is this:
you really, really should use the `phone` analyzer as the normal analyzer and `phone-search` as the search analyzer. deviating from this doesn't make a lot of sense as using the ngram analyzer (which is the normal one, not the search one!) on the search input would make the search input match everything which begins the same way as the phone number, which isn't what you want

reta replied:
Got it, thank you for explaining

rursprung commented:
> @rursprung thank you, could you please resolve the conflicts?

done (let's hope it stays mergeable for longer this time!)

reta reviewed (approved):
LGTM, thanks @rursprung ! @msfroh anything left on your side?

msfroh reviewed (approved):
Looks good, thanks!

reta commented:
@rursprung apologies, mind please sending a manual backport to`2.x` branch? thank you

rursprung commented:
> @rursprung apologies, mind please sending a manual backport to`2.x` branch? thank you

no worries, done: #16187


i'm a big fan of having a changelog, but it's causing a lot of merge conflicts here 🙁 
i've seen another repo (don't remember which, but i have a vague feeling that it was even in the opensearch org?) which had a subfolder where you created one file per PR for the changelog and some automated tooling then collected all of that together and merged it into the main changelog file for the release (and deleted the other files). maybe that might be an idea here as well to avoid the merge conflicts? might be worth discussing in a separate issue (which should probably come from someone regularly contributing to this repo as you'll be much more affected by this issue than me)?

on another note: squash-merging destroys my nice atomic commits 🙁 
i get that you do that for PRs where people just add a ton of "fix review finding" commits, but for the proper (linux kernel style ;)) PRs where you force-push to have a nice linear commit history with atomic commits, each with a nice commit message) i think this is a big loss (and makes it harder to find the actual culprit with tools like `git-bisect` and `git-revert`)

reta commented:
> might be worth discussing in a separate issue (which should probably come from someone regularly contributing to this repo as you'll be much more affected by this issue than me)?

Please feel free to open an issue or kick off discussion! 

> on another note: squash-merging destroys my nice atomic commits 

The clean repo history is useful, but this is a tradeoff for sure
